File: Data Structures & Algorithms/count-connected-components/submission-0.cpp
class Solution {
public:
    vector<int> parent;
    int find(int i){
        if(i==parent[i])
            return i;
        return i=find(parent[i]);
    }

    bool union_find(int a,int b){
        a=find(parent[a]);
        b=find(parent[b]);
        if(a==b)
            return true;
        parent[b]=a;
        return false; 
    }
    int countComponents(int n, vector<vector<int>>& edges) {
        parent.resize(n);
        for(int i=0;i<n;i++)
            parent[i]=i;
        int count=n;
        for(auto x:edges){
            if(!union_find(x[0],x[1]))
               count--;
        }
        
        
    return count;
    }
};

File: Data Structures & Algorithms/redundant-connection/submission-0.py



class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n=len(edges)
        parent=[i for i in range(n+1)]
        rank = [1]*(n+1)

        def find(i):
            while i != parent[i]:
                parent[i] = parent[parent[i]]
                i = parent[i]
            return i

        def union(x,y):
            p1=find(x)
            p2=find(y)
            if p1 == p2:
                return False
            if rank[p1]>rank[p2]:
                parent[p2]=p1
                rank[p1]+=rank[p2]
            else:
                parent[p1]=p2
                rank[p2]+=rank[p1]
            return True
        for x,y in edges:
            if not union (x,y):
                return [x,y]
            